baselines/FedSMOO/FedSMOO/main.py

```diff
@@ -0,0 +1,176 @@
+        num_rounds=cfg.num_rounds,
+        num_clients=cfg.num_clients,) 
+
+    else:
+        print("write a suitable message")
+        client_fn = None
+    
+    #----------------------------------------------------------------------------------------------
+
+    # 4. Define your strategy
+    # pass all relevant argument (including the global dataset used after aggregation,
+    # if needed by your method.)
+
+    # testloader = None
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    evaluate_fn = gen_evaluate_fn(testloader, device, cfg.model) 
+    
+    strategy = instantiate(cfg.strategy,
+                           evaluate_fn=evaluate_fn)
+
+    
+    #-----------------------------------------------------------------------------------------------
+    # 4.5 Create a custom server?
+
+    client_manager = SimpleClientManager()
+    
+    if cfg.method == "FedSMOO": 
+        server = instantiate(cfg.server,
+                        net = instantiate(cfg.model).to(device),
+                        sam_lr = cfg.sam_lr,
+                        client_manager = client_manager,
+                        strategy = strategy)
+    
+    elif cfg.method == "FedAvg":
+        server = Server(client_manager = client_manager,
+                        strategy = strategy)
+    
+    else:
+        # use strategy to initialize the default server
+        print("write a suitable message")
+        server = None
+        
+    
+    #-----------------------------------------------------------------------------------------------
+    # 5. Start Simulation
+    
+    print("\n---------------------- starting simulation --------------------------\n")
+
+    history = fl.simulation.start_simulation(
+                    client_fn=client_fn,
+                    num_clients=cfg.num_clients,
+                    server=server,
+                    config=fl.server.ServerConfig(num_rounds=cfg.num_rounds),
+                    client_resources={"num_cpus": cfg.client_resources.num_cpus,
+                                    "num_gpus": cfg.client_resources.num_gpus,},
+                    strategy=strategy)
+    #------------------------------------------------------------------------------------------------
+
+    # 6. Save your results
+    # Here you can save the `history` returned by the simulation and include
+    # also other buffers, statistics, info needed to be saved in order to later
+    # on generate the plots you provide in the README.md. You can for instance
+    # access elements that belong to the strategy for example:
+    # data = strategy.get_my_custom_data() -- assuming you have such method defined.
+    # Hydra will generate for you a directory each time you run the code. You
+    # can retrieve the path to that directory with this:
+    # save_path = HydraConfig.get().runtime.output_dir
+
+    accuracy_test = history.metrics_centralized["accuracy"]
+    plot_fn(cfg, accuracy_test)
```

Comment:
This is currently crashing because `./results` (hardcoded inside that function) doesn't exist. You can easily get the path for the experiment directory that's automatically generated by `Hydra` via:

```python
from hydra.core.hydra_config import HydraConfig
save_path = HydraConfig.get().runtime.output_dir
```

As you have probably noticed, a directory in `outputs/<current-date>/<current-time>` is generated each time you run your `main.py`. the `save_path` above will return you the path to that directory. Use that to save results and plots